File: BaekJoon/10026.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <tuple>
#include <functional>

using namespace std;
// 6 4 8
vector<vector<char>> board;
vector<vector<bool>> visited;

int dy[4] = {1, 0, -1, 0};
int dx[4] = {0, -1, 0, 1};
int n;

vector<function<bool(char, char)>> f = {
    [](char a, char b)
    {
        return a == b;
    },
    [](char a, char b)
    {
        if ((a == 'G' && b == 'R') || (a == 'R' && b == 'G'))
        {
            return true;
        }
        return a == b;
    }};

void dfs(int y, int x, int ii)
{
    visited[y][x] = true;
    for (int i = 0; i < 4; i++)
    {
        int yy = y + dy[i];
        int xx = x + dx[i];
        if (xx < 0 || yy < 0 || yy >= n || xx >= n || visited[yy][xx])
        {
            continue;
        }
        if (!f[ii](board[y][x], board[yy][xx]))
        {
            continue;
        }
        dfs(yy, xx, ii);
    }
}

void dfsAll()
{
    int aa = 0;
    visited.resize(n, vector<bool>(n, false));
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            if (!visited[i][j])
            {
                dfs(i, j, 0);
                aa++;
            }
        }
    }
    cout << aa << " ";
    aa = 0;
    visited.clear();
    visited.resize(n, vector<bool>(n, false));
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            if (!visited[i][j] == 1)
            {
                dfs(i, j, 1);
                aa++;
            }
        }
    }
    cout << aa;
}

int main()
{
    cin >> n;
    board.resize(n, vector<char>(n, 0));
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            cin >> board[i][j];
        }
    }
    dfsAll();
}

File: BaekJoon/_10026.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.*;

class _10026 {
	static final BufferedWriter out = new BufferedWriter(new OutputStreamWriter(System.out));
	static final BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
	static char board[][];
	static int visited[][];
	static int dx[] = { 0, 0, 1, -1 };
	static int dy[] = { 1, -1, 0, 0 };
	static HashMap<Character, Character> rgbMap = new HashMap<Character, Character>() {
		{
			put('B', 'B');
			put('R', 'G');
			put('G', 'R');
		}
	};

	public static void dfs(int y, int x, char mark, boolean isRedGreen) {
		visited[y][x] = mark;
		for (int i = 0; i < 4; i++) {
			int ny = y + dy[i];
			int nx = x + dx[i];
			if (nx == -1 || nx == board[0].length || ny == -1 || ny == board.length || visited[ny][nx] != 0) {
				continue;
			}
			if (board[ny][nx] == mark || (isRedGreen && rgbMap.get(board[ny][nx]) == mark)) {
				dfs(ny, nx, mark, isRedGreen);
			}
		}
	}

	public static int dfsAll(boolean isRedGreen) {
		int answer = 0;
		visited = new int[board.length][board[0].length];
		for (int i = 0; i < board.length; i++) {
			for (int j = 0; j < board[0].length; j++) {
				if (visited[i][j] == 0) {
					dfs(i, j, board[i][j], isRedGreen);
					answer++;
				}
			}
		}
		return answer;
	}

	public static void main(String args[]) throws Exception {
		int N = Integer.parseInt(in.readLine());

		board = new char[N][N];
		for (int i = 0; i < N; i++) {
			String s = in.readLine();
			for (int j = 0; j < N; j++) {
				board[i][j] = s.charAt(j);
			}
		}
		System.out.print(dfsAll(false));
		System.out.print(" ");
		System.out.print(dfsAll(true));
		out.flush();
		out.close();
		in.close();
	}
}
